From ae1be556de1679934fc8164f3b45b4c4ec8aed80 Mon Sep 17 00:00:00 2001
From: Houffaneh Osman <hosman@qti.qualcomm.com>
Date: Wed, 12 May 2021 08:12:50 +0200
Subject: [PATCH 1/1] vrend: Disable dual_src_blend feature
 (GL_EXT_blend_func_extended) for Qualcomm.

This patch file is from Qualcomm Technologies, Inc.  and is provided under the following license:

Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
SPDX-License-Identifier: BSD-3-Clause-Clear

The dual_src_blend feature as in support for GL_EXT_blend_func_extended
seems to be broken on Qualcomm Adreno. Reports from the past about broken
MRT (multi render target) support can be found on the internet as well,
so it looks like that didn't get fixed, yet.

The error messages happen at the linking stage, and follow a pattern similar
to this: "Error: Output ... location or component exceeds max allowed."
where the output variable is part of the multiple outputs from the fragment
shader stage.

With this change the feature is suppressed even if the driver reports
GL_EXT_blend_func_extended being available. Being an optional feature,
virglrenderer will adapt accordingly. This fix can also be used on
other platforms, however, it may limit their capabilities this way.
Exynos/Mali doesn't support this extension, for example, so it is safe
to use there.

Note: This is a workaround for a OpenGL/ES driver issue.

Issue: COQOS-8161
---
 src/vrend_renderer.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/vrend_renderer.c b/src/vrend_renderer.c
index be68f10..ebe2ee7 100644
--- a/src/vrend_renderer.c
+++ b/src/vrend_renderer.c
@@ -363,7 +363,7 @@ static inline bool has_feature(enum features_id feature_id)
    VREND_DEBUG(dbg_feature_use, NULL, "Try using feature %s:%d\n",
                feature_list[feature_id].log_name,
                vrend_state.features[feature_id]);
-   return (feature_id != feat_texture_view) && vrend_state.features[feature_id];
+   return (feature_id != feat_texture_view) && (feature_id != feat_dual_src_blend) && vrend_state.features[feature_id];
 }
 
 static inline void set_feature(enum features_id feature_id)
-- 
2.31.1

